### ISSUE TITLE
Preview on data table v4.0

### ISSUE THREAD

**Issue by rUriarteDev** [Laravel-Backpack/CRUD#2051]
When I enter the preview of a row in the data table, which has two FK attributes, I see the IDs of those FKs. But I need to see other attributes like when I edit the object. This happens only in this case that I have the FK as attributes.
View of datatable
![image](https://user-images.githubusercontent.com/39219144/64832505-9c9ce880-d5b0-11e9-9e14-2bfb899745c1.png)
View when click on edit
![image](https://user-images.githubusercontent.com/39219144/64832514-a4f52380-d5b0-11e9-8ca0-0d3192fb574b.png)
View when click on preview
![image](https://user-images.githubusercontent.com/39219144/64832525-ae7e8b80-d5b0-11e9-8513-e9268d910e28.png)

Any idea how to make it not show the FK and show the attributes like when I edit.

**Comment by rodrigouriarte05** [comment]
```
<?php

namespace App\Http\Controllers\Admin;

use App\Http\Requests\ComensalRequest;
use Backpack\CRUD\app\Http\Controllers\CrudController;
use Backpack\CRUD\app\Library\CrudPanel\CrudPanelFacade as CRUD;

/**
 * Class ComensalCrudController
 * @package App\Http\Controllers\Admin
 * @property-read CrudPanel $crud
 */
class ComensalCrudController extends CrudController
{
    use \Backpack\CRUD\app\Http\Controllers\Operations\ListOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\CreateOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\UpdateOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\DeleteOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\ShowOperation;

    public function setup()
    {
        $this->crud->setModel('App\Models\Comensal');
        $this->crud->setRoute(config('backpack.base.route_prefix') . '/comensal');
        $this->crud->setEntityNameStrings('comensal', 'comensales');
    }

    protected function setupListOperation()
    {
        $this->crud->addColumns(['usuario','comedor']);

        $this->crud->setColumnDetails('usuario', [
            'label' => 'Usuario',
            'type' => 'select',
            'name' => 'user_id', // the db column for the foreign key
            'entity' => 'user', // the method that defines the relationship in your Model
            'attribute' => 'name', // foreign key attribute that is shown to user
            'model' => "App\Models\BackpackUser" // foreign key model
        ]);

        $this->crud->setColumnDetails('comedor', [
            'label' => 'Comedor',
            'type' => 'select',
            'name' => 'comedor_id', // the db column for the foreign key
            'entity' => 'comedor', // the method that defines the relationship in your Model
            'attribute' => 'descripcion', // foreign key attribute that is shown to user
            'model' => "App\Models\Comedor" // foreign key model
        ]);
    }

    protected function setupCreateOperation()
    {
        $this->crud->setValidation(ComensalRequest::class);

        $this->crud->addField([  // Select2
            'label' => "Usuario",
            'type' => 'select2',
            'name' => 'user_id', // the db column for the foreign key
            'entity' => 'user', // the method that defines the relationship in your Model
            'attribute' => 'name', // foreign key attribute that is shown to user
            'model' => "App\Models\BackpackUser", // foreign key model

            // optional
            'default' => 1, // set the default value of the select2
            // 'options'   => (function ($query) {
            // return $query->orderBy('name', 'ASC')->where('depth', 1)->get();
            // }), // force the related options to be a custom query, instead of all(); you can use this to filter the results show in the select
        ]);

        $this->crud->addField([  // Select2
            'label' => "Comedor",
            'type' => 'select2',
            'name' => 'comedor_id', // the db column for the foreign key
            'entity' => 'comedor', // the method that defines the relationship in your Model
            'attribute' => 'descripcion', // foreign key attribute that is shown to user
            'model' => "App\Models\Comedor", // foreign key model

            // optional
            'default' => 1, // set the default value of the select2

            //'options'   => (function ($query) {
            //return $query->orderBy('name', 'ASC')->where('depth', 1)->get();
            //}), // force the related options to be a custom query, instead of all(); you can use this to filter the results show in the select
        ]);
    }

    protected function setupUpdateOperation()
    {
        $this->setupCreateOperation();
    }
}
```
I have no setupShowOperation() created by default. Do i have to create that function and write the addcolumn there. What are the differences with setupListOperation(). Thanks mate

**Comment by rodrigouriarte05** [comment]
Just read the doc, and correct it. Now it shows the other attributes, but still shows the FKs values, they are declared as fillable in the model. There is a wat to hidde them?
![image](https://user-images.githubusercontent.com/39219144/64864688-11524000-d60e-11e9-887b-94e5266c713f.png)

**Comment by rodrigouriarte05** [comment]
Im stucked with this, cant hide the attributes declared as fillable in the model. And if i remove them from the model they are still appearing in the preview.

**Comment by tabacitu** [comment]
@rodrigouriarte05 - you seem to have configured the List operation, yes, but the Show operation isn't configured. If you add this to your Controller:

```php
private function setupShowOperation()
{
    $this->setupListOperation();
}
```

does it fix it for you?

**Comment by rodrigouriarte05** [comment]
> @rodrigouriarte05 - you seem to have configured the List operation, yes, but the Show operation isn't configured. If you add this to your Controller:
> 
> ```
> private function setupShowOperation()
> {
>     $this->setupListOperation();
> }
> ```
> 
> does it fix it for you?

Actually i have already do that. But on preview mode still showing "User" and "Comedor" FKs
![image](https://user-images.githubusercontent.com/39219144/65151871-52c36080-d9fd-11e9-911a-69aaaa84ca21.png)
On List view it's OK not showing up the FKs
![image](https://user-images.githubusercontent.com/39219144/65151971-771f3d00-d9fd-11e9-949e-268ab520a165.png)
Here is the code of my controller

```
<?php

namespace App\Http\Controllers\Admin;

use App\Http\Requests\ComensalRequest;
use Backpack\CRUD\app\Http\Controllers\CrudController;
use Backpack\CRUD\app\Library\CrudPanel\CrudPanelFacade as CRUD;

/**
 * Class ComensalCrudController
 * @package App\Http\Controllers\Admin
 * @property-read CrudPanel $crud
 */
class ComensalCrudController extends CrudController
{
    use \Backpack\CRUD\app\Http\Controllers\Operations\ListOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\CreateOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\UpdateOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\DeleteOperation;
    use \Backpack\CRUD\app\Http\Controllers\Operations\ShowOperation;

    public function setup()
    {
        $this->crud->setModel('App\Models\Comensal');
        $this->crud->setRoute(config('backpack.base.route_prefix') . '/comensal');
        $this->crud->setEntityNameStrings('comensal', 'comensales');
    }

    protected function setupListOperation()
    {
        $this->crud->addColumns(['u', 'c']);

        $this->crud->setColumnDetails('u', [
            'label' => 'Usuario',
            'type' => 'select',
            'name' => 'user_id', // the db column for the foreign key
            'entity' => 'user', // the method that defines the relationship in your Model
            'attribute' => 'name', // foreign key attribute that is shown to user
            'model' => "App\Models\BackpackUser" // foreign key model
        ]);

        $this->crud->setColumnDetails('c', [
            'label' => 'Comedor',
            'type' => 'select',
            'name' => 'comedor_id', // the db column for the foreign key
            'entity' => 'comedor', // the method that defines the relationship in your Model
            'attribute' => 'descripcion', // foreign key attribute that is shown to user
            'model' => "App\Models\Comedor" // foreign key model
        ]);
    }

    protected function setupShowOperation()
    {
        $this->setupListOperation();   
    }

    protected function setupCreateOperation()
    {
        $this->crud->setValidation(ComensalRequest::class);

        $this->crud->addField([  // Select2
            'label' => "Usuario",
            'type' => 'select2',
            'name' => 'user_id', // the db column for the foreign key
            'entity' => 'user', // the method that defines the relationship in your Model
            'attribute' => 'name', // foreign key attribute that is shown to user
            'model' => "App\Models\BackpackUser", // foreign key model

            // optional
            'default' => 1, // set the default value of the select2
            // 'options'   => (function ($query) {
            // return $query->orderBy('name', 'ASC')->where('depth', 1)->get();
            // }), // force the related options to be a custom query, instead of all(); you can use this to filter the results show in the select
        ]);

        $this->crud->addField([  // Select2
            'label' => "Comedor",
            'type' => 'select2',
            'name' => 'comedor_id', // the db column for the foreign key
            'entity' => 'comedor', // the method that defines the relationship in your Model
            'attribute' => 'descripcion', // foreign key attribute that is shown to user
            'model' => "App\Models\Comedor", // foreign key model

            // optional
            'default' => 1, // set the default value of the select2

            //'options'   => (function ($query) {
            //return $query->orderBy('name', 'ASC')->where('depth', 1)->get();
            //}), // force the related options to be a custom query, instead of all(); you can use this to filter the results show in the select
        ]);
    }

    protected function setupUpdateOperation()
    {
        $this->setupCreateOperation();
    }
}
```

**Comment by tabacitu** [comment]
Well that's _very_ weird 😄 I've just tried it myself - both with addColumn and addColumn+setColumnDetails as you did, and... it worked:

![Screenshot 2019-09-18 at 17 24 45](https://user-images.githubusercontent.com/1032474/65157422-642a5e00-da39-11e9-8d86-2c9e2163081d.png)


If you copy your addColumn statements inside ```setupShowOperation()```... does it show the relation names then _then_?

**Comment by rodrigouriarte05** [comment]
> Well that's _very_ weird 😄 I've just tried it myself - both with addColumn and addColumn+setColumnDetails as you did, and... it worked:
> 
> ![Screenshot 2019-09-18 at 17 24 45](https://user-images.githubusercontent.com/1032474/65157422-642a5e00-da39-11e9-8d86-2c9e2163081d.png)
> 
> If you copy your addColumn statements inside `setupShowOperation()`... does it show the relation names then _then_?

The relations name are showed ok as i call setupListOperation(). But i dont want to show the foreign keys values of both relations.
I think the key is here in the documentation, im just dont understanding how to do it.
`This view uses the same column types you've defined in setup() and adds all other fillable attributes on the model. If you need to hide a ListEntries column from the Show operation, you can specify 'visibleInShow' => false, to the column in setup(), and it will be hidden from the preview page.`

**Comment by tabacitu** [comment]
Oh ok. You're right, it looks like the docs haven't been properly updated from 3.x to 4.0. I've ran through the docs to make it more clear how it works in 4.0

I'm sorry but I still don't understand what you're trying to do. Aren't you trying to show the "u" and "c" relationships in the Show operation (by showing their name/descripcion attributes, not ID)?

**Comment by rodrigouriarte05** [comment]
> Oh ok. You're right, it looks like the docs haven't been properly updated from 3.x to 4.0. I've ran through the docs to make it more clear how it works in 4.0
> 
> I'm sorry but I still don't understand what you're trying to do. Aren't you trying to show the "u" and "c" relationships in the Show operation (by showing their name/descripcion attributes, not ID)?

Exactly, im trying to show their name/description attributes. But get both the name/description and the ID's which I dont know why are showing if hey are not declared.

**Comment by rodrigouriarte05** [comment]
Ok this was really strange, maybe is an issue or just an exceptional case.
To use operation in CRUD we import operations from 

> \Backpack\CRUD\app\Http\Controllers\Operations

 In my project that folder does not exist and the Operations are in

> vendor\backpack\crud\src\app\Http\Controllers\Operations\ShowOperation.php

Is this normal? Or I did a wrong install of Backpack?

And inside that exist `public function show($id)` where is declared `$this->crud->setFromDb();`
If I null that line, ID's taken from DB are not shown anymore.

This line you added in the documentation 
`$this->crud->set('show.setFromDb', false);`
I think is not working, otherwise should not show the DB attributes.

**Comment by tabacitu** [comment]
Yes it's normal. ```\Backpack\CRUD\app\ == vendor\backpack\crud\src\app\```

Oh... I think I get it now. To sum up:
- you've defined the columns in ```setupListOperation()```;
- ran ```setupListOperation()``` inside ```setupShowOperation()```;
- then ```setFromDb()``` is _overwriting_ the relationship columns you defined in ```setupListOperation()```, showing IDs instead of attributes;

### Solution 1

Please do a ```composer update``` (I just pushed the ```show.setFromDb``` flag yesterday), then  ```$this->crud->set('show.setFromDb', false);``` first thing inside your ```setupShowOperation()```. That should fix it for you. But then it'll show the same columns in List and Show.

### Solution 2

That root cause must be because of the NAMES of the columns. I see you've defined them in a pretty weird way - first ```$this->crud->addColumns(['u', 'c']);``` then renaming them to their proper name with ```setColumnDetails()``` (the column in the db). 

Try using addColumn instead:
```php
    protected function setupListOperation()
    {
        $this->crud->addColumn([
            'label' => 'Usuario',
            'type' => 'select',
            'name' => 'user_id', // the db column for the foreign key
            'entity' => 'user', // the method that defines the relationship in your Model
            'attribute' => 'name', // foreign key attribute that is shown to user
            'model' => "App\Models\BackpackUser" // foreign key model
        ]);

        $this->crud->addColumn([
            'label' => 'Comedor',
            'type' => 'select',
            'name' => 'comedor_id', // the db column for the foreign key
            'entity' => 'comedor', // the method that defines the relationship in your Model
            'attribute' => 'descripcion', // foreign key attribute that is shown to user
            'model' => "App\Models\Comedor" // foreign key model
        ]);
    }
```
Or using the same names with addColumns as in setColumnDetails (```comedor_id``` and ```user_id```).

I suspect that might be the problem. And you won't need the workaround above.

Cheers!

**Comment by DanielBalsalobre** [comment]
If you experience this problem try to redefine the column in setupShowOperation whitout the 'model' parameter

**Comment by mahasadhu** [comment]
@DanielBalsalobre you just saved my day! My modifyColumn function was not working in show, but working in list operation

when I remove the model parameter from array, it worked like a charm

**Comment by gstrat88** [comment]
app/Http/Controllers/Operations/ShowOperation.php line 90 for some reason it removes columns with model set and autoset = true

so if someone uses modifyColumn to add the relation he should add the option 'autoset' => false to bypass the check

**Comment by mahasadhu** [comment]
@gstrat88 ah yes I see..

@tabacitu @pxpm could you please elaborate ? Is this intended or an actual error ?

**Comment by tabacitu** [comment]
@mahasadhu it's intended, but... it's not intuitive at all, and we plan to change the default behaviour in the next version of Backpack. Please see my answer in https://github.com/Laravel-Backpack/CRUD/issues/3108#issuecomment-676425042 which explains it in more detail, and PR https://github.com/Laravel-Backpack/CRUD/pull/3132 which hopefully fixes this once and for all 😌 by providing `setupShowOperation()` to _completely overwrite_ which columns to show, but also `modifyShowOperation()` that will allow you to just add/remove/modify columns in the default behaviour. 

If you could give feedback on PR https://github.com/Laravel-Backpack/CRUD/pull/3132 that would be great, I'd appreciate it. I think it will solve most problems people had with the Show operation, which indeed has been quite unintuitive until now.

I'm going to close this as we have a fix for it in PR https://github.com/Laravel-Backpack/CRUD/pull/3132, even though we won't be able to merge the fix until sep-oct 2020, when Backpack 4.2 is launched.

Cheers! Thanks everyone for your patience on this.
